Geometry_Engine/Modify/Offset.cs

```diff
@@ -0,0 +1,868 @@
+                else
+                    result.Add((curve as Line).Extend(start, end, false, tolerance));
+            }
+            else
+            {
+                if (curve is Arc)
+                {
+                    if (start >= 0 && end >= 0)
+                        result.Add((curve as Arc).Extend(start, end, true, tolerance));
+                    else if (start >= 0 && end < 0)
+                        result.Add((curve as Arc).Extend(start, 2 * Math.Asin(end / (2 * (curve as Arc).Radius)) * (curve as Arc).Radius, true, tolerance));
+                    else if (start < 0 && end >= 0)
+                        result.Add((curve as Arc).Extend(2 * Math.Asin(start / (2 * (curve as Arc).Radius)) * (curve as Arc).Radius, end, true, tolerance));
+                    else
+                        result.Add((curve as Arc).Extend(2 * Math.Asin(start / (2 * (curve as Arc).Radius)) * (curve as Arc).Radius, 2 * Math.Asin(end / (2 * (curve as Arc).Radius)) * (curve as Arc).Radius, true, tolerance));
+                }
+                else
+                    result.Add((curve as Line).Extend(start, end, true, tolerance));
+            }
+
+            for (int i = 0; i < result.Count; i++)
+                if (result[i] is PolyCurve)
+                {
+                    PolyCurve temp = (PolyCurve)result[i];
+                    result.RemoveAt(i);
+                    result.InsertRange(i, temp.Curves);
+                }
+
+            return result;
+        }
+
+        /***************************************************/
+
+        private static PolyCurve Fillet(this ICurve curve1, ICurve curve2, bool tangentExtensions, bool keepCurve1StartPoint, bool keepCurve2StartPoint, double tolerance = Tolerance.Distance)
+        {
+            //TODO:
+            //Write a proper fillet method, test and make it public            
+
+            if (!((curve1 is Line || curve1 is Arc) && (curve2 is Line || curve2 is Arc))) //for now works only with combinations of lines and arcs
+                throw new NotImplementedException();
+
+            List<ICurve> resultCurves = new List<ICurve>();
+
+            bool C1SP = keepCurve1StartPoint;
+            bool C2SP = keepCurve2StartPoint;
+
+            List<Point> intersections = curve1.ICurveIntersections(curve2, tolerance);
+            if (intersections.Count > 2)
+                Reflection.Compute.RecordError("How on Earth did you manage to create a pair of Arcs and/or Lines which have more than two intersections??");
+            else if (intersections.Count == 2 || intersections.Count == 1)
+            {
+                Point intersection = intersections[0];
+                if (intersections.Count == 2)
+                {
+                    double maxdist = double.MaxValue;
+                    if (C1SP)
+                        foreach (Point p in intersections)
+                        {
+                            if (p.Distance(curve1.IStartPoint()) < maxdist)
```

Comment:
`SquareDistance` could be used instead of `Distance` plus please declare store the result as variable to avoid overhead 2 lines later (same refers to the next loop)